ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/vm/VmTraceGenerator.java

```diff
@@ -0,0 +1,284 @@
+      final VmTrace newSubTrace = new VmTrace();
+      parentTraces.addLast(newSubTrace);
+      findLastFrameInCall(currentTraceFrame, currentIndex)
+          .ifPresent(
+              lastFrameInCall -> {
+                report.setUsed(lastFrameInCall.getGasRemaining().toLong());
+                lastFrameInCall
+                    .getStack()
+                    .filter(stack -> stack.length > 0)
+                    .map(stack -> stack[stack.length - 1])
+                    .map(last -> Quantity.create(UInt256.fromHexString(last.getHexString())))
+                    .ifPresent(report::addPush);
+              });
+      op.setSub(newSubTrace);
+    }
+  }
+
+  private void handleDepthDecreased() {
+    // check if next frame depth has decreased i.e the current operation closes the parent trace
+    if (maybeNextFrame.map(next -> next.isLessDeepThan(currentTraceFrame)).orElse(false)) {
+      currentTrace = parentTraces.removeLast();
+    }
+  }
+
+  private VmOperation buildVmOperation() {
+    final VmOperation op = new VmOperation();
+    // set gas cost and program counter
+    op.setCost(currentTraceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+    op.setPc(currentTraceFrame.getPc());
+    return op;
+  }
+
+  private VmOperationExecutionReport generateExecutionReport() {
+    final VmOperationExecutionReport report = new VmOperationExecutionReport();
+    // set gas remaining
+    report.setUsed(
+        currentTraceFrame.getGasRemaining().toLong()
+            - currentTraceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+    return report;
+  }
+
+  private void generateTracingMemory(final VmOperationExecutionReport report) {
+    if (currentTraceFrame.getDepth() == maybeNextFrame.map(TraceFrame::getDepth).orElse(0)) {
+      updatedMemory(currentTraceFrame.getMemory(), maybeNextFrame.flatMap(TraceFrame::getMemory))
+          .filter(memory -> memory.length > 0)
+          .map(TracingUtils::dumpMemoryAndTrimTrailingZeros)
+          .map(Mem::new)
+          .ifPresent(report::setMem);
+    }
+  }
+
+  private void generateTracingPush(final VmOperationExecutionReport report) {
+    // set push from stack elements if some elements have been produced
+    if (currentTraceFrame.getStackItemsProduced() > 0 && maybeNextFrame.isPresent()) {
+      final Bytes32[] stack = maybeNextFrame.get().getStack().orElseThrow();
+      if (stack.length > 0) {
+        IntStream.range(0, currentTraceFrame.getStackItemsProduced())
+            .forEach(
+                i -> {
+                  final BytesValue value =
+                      BytesValues.trimLeadingZeros(stack[stack.length - i - 1]);
+                  report.addPush(Quantity.create(UInt256.fromHexString(value.getHexString())));
+                });
+      }
+    }
+  }
+
+  private void generateTracingStorage(final VmOperationExecutionReport report) {
+    // set storage if updated
+    updatedStorage(lastCapturedStorage, currentTraceFrame.getStorage())
```

Comment:
What happens if storage from two different accounts are modified?  It looks like this code assumes all storage belongs to the same account.  Can we add tests for this?

Reply:
Is it possible from an opcode level trace frame to another ?

Reply:
A single tx can make message calls to various contracts, so it's not guaranteed that all storage changes belong to the same account.

Reply:
i agree but here it is opcode level. a single vm operation can't update multiple account right ?